File: src/utils/storage/CloudDataStorage.ts
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { DataStorage } from "./declare/DataStorage";

export function CloudDataStorage<SchemaStorage extends object>(defaultValues: SchemaStorage, storageName = 'storage'): DataStorage<SchemaStorage> {
    type Keys = keyof typeof defaultValues;
    const cache = { ...defaultValues };
    const prefix = storageName + '_';

    const storage = Ads.bridge.storage;
    /* jstodef.add_listener((_, message_id, message) => {
        pprint(message_id)
        pprint(message)
    }) */
    load();

    function set<T extends Keys>(key: T, value: SchemaStorage[T]) {
        pprint('save to cloud storage');
        cache[key] = value;
        storage.set(prefix + (key as string), tostring(value));
    }

    function get<T extends Keys>(key: T): SchemaStorage[T] {
        return cache[key];
    }

    function load() {
        Object.keys(defaultValues).forEach(key => _loadItem(key));
    }

    function _loadItem(key: string) {
        const val = storage.get(prefix + key);
        pprint(storage.get(prefix + key));
        if (val == undefined || val == '')
            set(key as Keys, defaultValues[key as Keys]);
        else
            cache[key as Keys] = json.decode(val, {}).value;
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-argument



    }

    function clear() {
        Object.keys(cache).forEach(key => {
            set(key as Keys, defaultValues[key as Keys]);
        });
    }

    return {
        set, get, load, clear, schemaObject: defaultValues
    };
}

File: app/utils/storage/CloudDataStorage.lua
local ____lualib = require("lualib_bundle")
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__ObjectKeys = ____lualib.__TS__ObjectKeys
local __TS__ArrayForEach = ____lualib.__TS__ArrayForEach
local ____exports = {}
function ____exports.CloudDataStorage(defaultValues, storageName)
    if storageName == nil then
        storageName = "storage"
    end
    local set, load, _loadItem, cache, prefix, storage
    function set(key, value)
        pprint("save to cloud storage")
        cache[key] = value
        storage.set(
            prefix .. key,
            tostring(value)
        )
    end
    function load()
        __TS__ArrayForEach(
            __TS__ObjectKeys(defaultValues),
            function(____, key) return _loadItem(key) end
        )
    end
    function _loadItem(key)
        local val = storage.get(prefix .. key)
        pprint(storage.get(prefix .. key))
        if val == nil or val == "" then
            set(key, defaultValues[key])
        else
            cache[key] = json.decode(val, {}).value
        end
    end
    cache = __TS__ObjectAssign({}, defaultValues)
    prefix = storageName .. "_"
    storage = Ads.bridge.storage
    load()
    local function get(key)
        return cache[key]
    end
    local function clear()
        __TS__ArrayForEach(
            __TS__ObjectKeys(cache),
            function(____, key)
                set(key, defaultValues[key])
            end
        )
    end
    return {
        set = set,
        get = get,
        load = load,
        clear = clear,
        schemaObject = defaultValues
    }
end
return ____exports
